File: src/main/kotlin/drivers/PRISMDriver.kt
package drivers

import fmgp.did.method.prism.*
import fmgp.did.method.prism.cardano.*
import fmgp.did.method.prism.vdr.*
import interfaces.Driver
import interfaces.Proof
import java.security.PublicKey
import java.util.UUID
import kotlin.emptyArray
import org.hyperledger.identus.apollo.derivation.HDKey
import org.hyperledger.identus.apollo.derivation.MnemonicHelper
import org.hyperledger.identus.apollo.utils.KMMECSecp256k1PrivateKey

/**
 * InMemoryDriver is an example implementation of the [Driver] interface. It stores data in a
 * mutable map in memory. This implementation is useful for testing or scenarios where persistent
 * storage is not required.
 *
 * @property storage A mutable map that holds the stored data.
 */
class PRISMDriver(
        val bfConfig: BlockfrastConfig,
        wallet: CardanoWalletConfig,
) : Driver {

  override val identifier: String = ""
  override val family: String = ""
  override val version: String = "1.0"
  override val supportedVersions: Array<String> = emptyArray()

  val workdir: String = "../../prism-vdr/mainnet"
  val didPrism: DIDPrism =
          DIDPrism("51d47b13393a7cc5c1afc47099dcbecccf0c8a70828c072ac82f55225b42d4f4")
  val keyName: String = "vdr1"

  val seed =
          MnemonicHelper.Companion.createSeed(
                  listOf(
                          "mention",
                          "side",
                          "album",
                          "physical",
                          "uncle",
                          "lab",
                          "horn",
                          "nasty",
                          "script",
                          "few",
                          "hazard",
                          "announce",
                          "upon",
                          "group",
                          "ten",
                          "moment",
                          "fantasy",
                          "helmet",
                          "supreme",
                          "early",
                          "gadget",
                          "curve",
                          "lecture",
                          "edge"
                  ),
                  ""
          )

  val vdrKey: KMMECSecp256k1PrivateKey = HDKey(seed, 0, 1).getKMMSecp256k1PrivateKey()

  private /*lateinit*/ var genericVDRDriver: GenericVDRDriver

  init {
    genericVDRDriver =
            GenericVDRDriver(
                    bfConfig,
                    wallet,
                    workdir, // : String = "../../prism-vdr/mainnet",
                    didPrism,
                    keyName,
                    vdrKey
            )

    genericVDRDriver.initState()
    // TODO
    // initialiseVDRIndexer()
  }

  fun test() = ""

  /** Exception thrown when data cannot be found in the storage. */
  class DataCouldNotBeFoundException : Exception("Could not find the data")

  var storage: MutableMap<String, ByteArray> = mutableMapOf()

  override fun create(data: ByteArray, options: Map<String, Any>?): Driver.OperationResult {
    val uuid = UUID.randomUUID().toString()
    storage += mutableMapOf(Pair(uuid, data))

    return Driver.OperationResult(
            UUID.randomUUID().toString(),
            Driver.OperationState.SUCCESS,
            emptyArray(),
            emptyMap(),
            uuid,
            null,
            null
    )
  }

  override fun update(
          data: ByteArray,
          paths: Array<String>,
          queries: Map<String, String>,
          fragment: String?,
          options: Map<String, Any>?
  ): Driver.OperationResult {
    if (fragment != null) {
      storage += mutableMapOf(Pair(fragment, data))

      return Driver.OperationResult(
              UUID.randomUUID().toString(),
              Driver.OperationState.SUCCESS,
              emptyArray(),
              emptyMap(),
              fragment,
              null,
              null
      )
    } else {
      throw DataCouldNotBeFoundException()
    }
  }

  override fun read(
          paths: Array<String>,
          queries: Map<String, String>,
          fragment: String?,
          publicKeys: Array<PublicKey>?
  ): ByteArray {
    if (fragment != null) {
      return storage.getValue(fragment)
    } else {
      throw DataCouldNotBeFoundException()
    }
  }

  override fun delete(
          paths: Array<String>,
          queries: Map<String, String>,
          fragment: String?,
          options: Map<String, Any>?
  ) {
    if (fragment != null) {
      storage.remove(fragment)
    } else {
      throw DataCouldNotBeFoundException()
    }
  }

  override fun verify(
          paths: Array<String>,
          queries: Map<String, String>,
          fragment: String?,
          publicKeys: Array<PublicKey>?,
          returnData: Boolean
  ): Proof {
    val dataRead = read(paths, queries, fragment, publicKeys)
    val sendData =
            if (returnData) {
              dataRead
            } else {
              null
            }
    return Proof(
            "SHA256",
            sendData,
            ByteArray(0) // Empty SHA256 for now we dont want to implement
    )
  }

  override fun storeResultState(identifier: String): Driver.OperationState {
    return Driver.OperationState.SUCCESS
  }
}

File: src/main/scala/hyperledger/identus/vdr/prism/PrismDriver.scala
package hyperledger.identus.vdr.prism

import scala.jdk.CollectionConverters.*
import fmgp.did.method.prism.*
import fmgp.did.method.prism.cardano.*
import fmgp.did.method.prism.vdr.*
import org.hyperledger.identus.apollo.utils.KMMECSecp256k1PrivateKey
import interfaces.Driver
import interfaces.Proof
import interfaces.Driver.OperationState

class DataCouldNotBeFoundException(reason: Option[String])
    extends Exception("Could not find the data" + reason.map(s => s" becuase: $s").getOrElse(""))
class DataNotInitializedException extends Exception("Data wasn't initialized")
class DataAlreadyDeactivatedException extends Exception("Data was deactivated")
//class InvalidRequest(reason: String) extends Exception(s"Invalid request because: $reason")

object PRISMDriver {

  // TODO those should come from our configuration file
  val didPrism: DIDPrism = DIDPrism("51d47b13393a7cc5c1afc47099dcbecccf0c8a70828c072ac82f55225b42d4f4")
  val keyName: String = "vdr1"
  private val wallet = CardanoWalletConfig(
    Seq(
      "mention",
      "side",
      "album",
      "physical",
      "uncle",
      "lab",
      "horn",
      "nasty",
      "script",
      "few",
      "hazard",
      "announce",
      "upon",
      "group",
      "ten",
      "moment",
      "fantasy",
      "helmet",
      "supreme",
      "early",
      "gadget",
      "curve",
      "lecture",
      "edge"
    )
  )

  private val vdrKey: KMMECSecp256k1PrivateKey = wallet.secp256k1PrivateKey(0, 1)

  def default: PRISMDriver = PRISMDriver(
    bfConfig = BlockfrostConfig(token = "preprod9EGSSMf6oWb81qoi8eW65iWaQuHJ1HwB"),
    wallet = wallet,
    didPrism = didPrism,
    vdrKey = vdrKey,
    keyName = "vdr1",
    workdir = "../../prism-vdr/mainnet",
  )
}

case class PRISMDriver(
    val bfConfig: BlockfrostConfig,
    wallet: CardanoWalletConfig,
    didPrism: DIDPrism,
    vdrKey: KMMECSecp256k1PrivateKey,
    keyName: String = "vdr1",
    workdir: String = "../../prism-vdr/mainnet"
) extends Driver {
  // implement methods here (or use stub for now)
  def getFamily: String = "PRISM"
  def getIdentifier: String = "PRISMDriver"
  def getVersion: String = "1.0"
  def getSupportedVersions: Array[String] = Array("1.0")

  private val genericVDRDriver: GenericVDRDriver =
    GenericVDRDriver(
      bfConfig,
      wallet,
      workdir = workdir,
      didPrism,
      keyName,
      vdrKey,
      maybeMsgCIP20 = Some(getIdentifier),
    )

  genericVDRDriver.initState
  // TODO check the status of the SSI and the key

  override def create(
      data: Array[Byte],
      options: java.util.Map[String, ?]
  ): interfaces.Driver.OperationResult = {
    GenericVDRDriver.runProgram(
      for {
        ret <- genericVDRDriver.createBytesEntry(data)
        out = Driver.OperationResult(
          ret._1.value,
          Driver.OperationState.SUCCESS,
          Array(ret._1.value),
          Map(("h", ret._1.value)).asJava,
          null,
          null,
          null
        )
      } yield out
    )
  }

  override def update(
      data: Array[Byte],
      paths: Array[String],
      queries: java.util.Map[String, String],
      fragment: String,
      options: java.util.Map[String, ?]
  ): interfaces.Driver.OperationResult = {
    paths.headOption match
      case None =>
        throw DataCouldNotBeFoundException(
          Some("the identifier is missing from the path")
        ) // interfaces.Driver.OperationResult
      case Some(hash) =>
        val eventRef: RefVDR = RefVDR(hash)
        GenericVDRDriver.runProgram(
          for {
            ret <- genericVDRDriver.updateBytesEntry(eventRef, data)
            (updateEventHash, code, string) = ret
            out = Driver.OperationResult(
              updateEventHash.hex,
              Driver.OperationState.SUCCESS,
              Array(eventRef.eventHash.hex),
              Map(("h", eventRef.eventHash.hex)).asJava,
              null,
              null,
              null
            )
          } yield out
        )
  }

  override def delete(
      paths: Array[String],
      queries: java.util.Map[String, String],
      fragment: String,
      options: java.util.Map[String, ?]
  ): Unit = {
    paths.headOption match
      case None =>
        throw DataCouldNotBeFoundException(
          Some("the identifier is missing from the path")
        ) // interfaces.Driver.OperationResult
      case Some(hash) =>
        val eventRef: RefVDR = RefVDR(hash)
        GenericVDRDriver.runProgram(
          for {
            ret <- genericVDRDriver.deleteBytesEntry(eventRef)
            (updateEventHash, code, string) = ret
            out = Driver.OperationResult(
              updateEventHash.hex,
              Driver.OperationState.SUCCESS,
              Array(eventRef.eventHash.hex),
              Map(("h", eventRef.eventHash.hex)).asJava,
              null,
              null,
              null
            )
          } yield out
        )
  }

  override def read(
      paths: Array[String],
      queries: java.util.Map[String, String],
      fragment: String,
      publicKeys: Array[java.security.PublicKey]
  ): Array[Byte] = {
    paths.headOption match
      case None => Array.empty()
      case Some(hash) =>
        val eventRef: RefVDR = RefVDR(hash)
        GenericVDRDriver.runProgram(
          for {
            vdr <- genericVDRDriver.read(eventRef)
          } yield vdr.data match {
            case VDR.DataEmpty()              => Array.empty()
            case VDR.DataDeactivated(data)    => Array.empty()
            case VDR.DataByteArray(byteArray) => byteArray
            case VDR.DataIPFS(cid)            => Array.empty()
            case VDR.DataStatusList(status)   => Array.empty()
          }
        )
  }

  /** in the case of PRISM this identifier is the HASH of the event */
  override def storeResultState(
      identifier: String
  ): interfaces.Driver.OperationState =
    OperationState.SUCCESS
    // OperationState.RUNNING //TODO since everything is a synchronous, the driver needs to keep a internal state for this
    // OperationState.ERROR //TODO only makes sense when the event submitted is not the latest version or if the submission didn't end up the the blockchain

  override def verify(
      paths: Array[String],
      queries: java.util.Map[String, String],
      fragment: String,
      publicKeys: Array[java.security.PublicKey],
      returnData: Boolean
  ): interfaces.Proof = {
    paths.headOption match
      case None => ???
      case Some(hash) =>
        val eventRef: RefVDR = RefVDR(hash)
        GenericVDRDriver.runProgram(
          for {
            vdr <- genericVDRDriver.read(eventRef)
          } yield vdr.data match {
            case VDR.DataEmpty() => Proof("PrismBlock", Array.empty(), Array.empty()) // TODO proof
            case VDR.DataDeactivated(data) =>
              data match {
                case VDR.DataEmpty()           => throw DataNotInitializedException()
                case VDR.DataDeactivated(data) => throw DataAlreadyDeactivatedException()
                case VDR.DataByteArray(byteArray) =>
                  Proof(
                    "PrismBlock",
                    byteArray, // Data
                    Array.empty() // TODO proof will is a protobuf Array of PRISM events. Reuse the PrismBlock?
                  )
                case VDR.DataIPFS(cid)          => ??? // not yet implemented
                case VDR.DataStatusList(status) => ??? // not yet implemented
              }
            case VDR.DataByteArray(byteArray) =>
              Proof(
                "PrismBlock",
                byteArray, // Data
                Array.empty() // TODO proof will is a protobuf Array of PRISM events like a PrismBlock?
              )
            case VDR.DataIPFS(cid)          => ??? // not yet implemented
            case VDR.DataStatusList(status) => ??? // not yet implemented
          }
        )
  }

}

object DemoPRISMDriver { // ./gradlew run
  def main(args: Array[String]): Unit = {
    println("RUN PRISMDriver")
    println(PRISMDriver.default.getVersion)
    val aux = PRISMDriver.default.create("Test PRISMDriver".getBytes, Map.empty.asJava)
    // https://preprod.cardanoscan.io/transaction/774d749d42243d27c308b5e849d3cf9848c05bcd95fa28a3f48f4cfeee1f33bc?tab=metadata
    println(aux) //
  }
}
